zk_ee/src/system/metadata.rs

```diff
@@ -71,12 +74,169 @@ impl UsizeDeserializable for BlockHashes {
     }
 }
 
+#[cfg_attr(feature = "testing", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Copy, Debug, PartialEq, Default)]
+pub struct InteropRoot {
+    pub root: Bytes32,
+    pub block_or_batch_number: u64,
+    pub chain_id: u64,
+}
+
+#[derive(Clone, Debug, PartialEq, Default)]
+pub struct InteropRootsContainer {
+    roots: ArrayVec<InteropRoot, MAX_NUMBER_INTEROP_ROOTS>,
+    length: u32,
+}
+
+impl From<ArrayVec<InteropRoot, MAX_NUMBER_INTEROP_ROOTS>> for InteropRootsContainer {
+    fn from(roots: ArrayVec<InteropRoot, MAX_NUMBER_INTEROP_ROOTS>) -> Self {
+        let length = roots.len().try_into().expect("Invalid amount of roots");
+        Self { roots, length }
+    }
+}
+
+impl From<InteropRootsContainer> for ArrayVec<InteropRoot, MAX_NUMBER_INTEROP_ROOTS> {
+    fn from(container: InteropRootsContainer) -> Self {
+        container.roots
+    }
+}
+
+impl InteropRootsContainer {
+    pub(crate) const EMPTY_VALUE: InteropRoot = InteropRoot {
+        root: Bytes32::ZERO,
+        block_or_batch_number: 0,
+        chain_id: 0,
+    };
+
+    pub fn roots(&self) -> &ArrayVec<InteropRoot, MAX_NUMBER_INTEROP_ROOTS> {
+        &self.roots
+    }
+}
+
+#[cfg(feature = "testing")]
+impl serde::Serialize for InteropRootsContainer {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.roots.to_vec().serialize(serializer)
+    }
+}
+
+#[cfg(feature = "testing")]
+impl<'de> serde::Deserialize<'de> for InteropRootsContainer {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let vec: Vec<InteropRoot> = Vec::deserialize(deserializer)?;
+        let mut array_vec = ArrayVec::new();
+        for item in vec {
+            if array_vec.try_push(item).is_err() {
+                return Err(serde::de::Error::custom(format!(
+                    "Too many InteropRoot items for ArrayVec (max {MAX_NUMBER_INTEROP_ROOTS})"
+                )));
+            }
+        }
+        let len = array_vec.len().try_into().expect("Deserialization failed");
+        Ok(Self {
+            roots: array_vec,
+            length: len,
+        })
+    }
+}
+
+impl UsizeSerializable for InteropRootsContainer {
```

Comment:
@AntonD3 pls take a look here. Maybe you'll have some better ideas how to encode/decode it. Iiuc, we can't have any dynamic structures here